controllers/ratelimitpolicy_limits.go

```diff
@@ -105,8 +107,38 @@ func (r *RateLimitPolicyReconciler) buildRateLimitIndex(ctx context.Context, rlp
 			return nil, err
 		}
 
+		if err := r.applyOverrides(ctx, rlp, targetNetworkObject); err != nil {
+			return nil, err
+		}
+
 		rateLimitIndex.Set(rlpKey, rlptools.LimitadorRateLimitsFromRLP(rlp))
 	}
 
 	return rateLimitIndex, nil
 }
+
+// applyOverrides checks for any overrides set for the RateLimitPolicy.
+// It iterates through the RateLimitPolicyList to find overrides for the provided target HTTPRoute.
+// If an override is found, it updates the limits in the RateLimitPolicySpec in accordingly.
+func (r *RateLimitPolicyReconciler) applyOverrides(ctx context.Context, rlp *kuadrantv1beta2.RateLimitPolicy, targetNetworkObject client.Object) error {
+	if route, ok := targetNetworkObject.(*gatewayapiv1.HTTPRoute); ok {
```

Comment:
I think we could follow a different approach here that is indifferent to the fact that nowadays we only allow one policy of a kind targeting a given network resource at a time.

Instead, what about the following?
1. List all the policies that target the same resource, plus, if it's a route policy, all the policies that target any of the route's parent Gateways
2. Sort the list of policies by precedence – i.e. the higher the target is in the [hierarchy](https://gateway-api.sigs.k8s.io/geps/gep-713/#hierarchy), the higher the precedence; same level in the hierarchy, older policy beats newer ones
3. Drop all policies with lower precedence than `rlp`
4. Iterate from highest to lowest until finding a block of overrides, in which case replace the `rlp`'s limits with it and `break`

In the future, for adding support for the `merge` strategy, we invert the order of the list (iterating from lowest to highest) and go all the way through the list (i.e. without breaking at the first match.)

A remaining problem to solve – which is an issue in your current proposed implementation also – is what to do in case of multiple parent gateways with override policies targeting them. I suppose we'll have to add the host names in.

WDYT?

Reply:
I am torn with my view on this. Yes the use of the DAG make it an easier problem. If code freeze is in two weeks I think it would be better to have a working solution sooner.

The conflict I have is with the hierarchy. Currently we only support polices attached to the gateway or the route. If the gateway has an atomic override the number of policies attached to route does not matter, the gateway always wins. 

The same goes for the multiply routes, we currently don't support that as a feature. While we possible should to me it is feeling like scope creep and solving a problem of tomorrow. A problem we are not sure even exists in the wild.

This is were my conflict is. I think we should be solving that problem, and that we should support polices attached to the gateway class. That is not the product of today, but of tomorrow.

Reply:
I see your point, @Boomatang. The part that I don't like is the spreading of one decision across the entire code, i.e. the coupling of implementations at multiple levels to a rule otherwise enforced at a single point, possibly at a level far beyond. We are just increasing the number of points in the code that will need to change when the rule also shifts.

In this case, the "decision" (or "rule") we'd be coupling the implementation to or not is the current behaviour of 1 policy of a kind per target. I already lost count of how many places in the code we rely on that for something. And this one is particularly nasty because the rule itself is not even obvious at a glance. Instead, one who reads the code needs to have that click that "oh, right! this is because of only 1 policy per..."

Other examples include for sure the kinds of resources that can be targeted by a policy. Dozens of ifs and switch cases all over for this one too, when IMO in many cases the types could/should abstract the complexity. (BTW, you can git blame me on several of those.)

As for using the DAG, I may be overlooking the difficulties to add it to implementation here, so I'll let you and @KevFan decide. My initial intuition is that it could result in some DRYer code and arguably improve performance.

Reply:
Hmm, yeah might play around with this, although part of this ,at least for RLP, might be addressed/implemented by @eguzki already since he has draft PR introducing a new limitador limits controller which uses the DAG https://github.com/Kuadrant/kuadrant-operator/pull/527

Reply:
Pushed an implememtation using the DAG that is focused solely on the overriden portion of the reconcile logic based on the algorithm suggested (https://github.com/Kuadrant/kuadrant-operator/pull/527 would be the general refactor to use the DAG throughout)

Reply:
Currently doesn't account for:

> A remaining problem to solve – which is an issue in your current proposed implementation also – is what to do in case of multiple parent gateways with override policies targeting them. I suppose we'll have to add the host names in.

In this scenario, the oldest Gateway RLP's override block is used